plugins/performance-lab/includes/admin/rest-api.php

```diff
@@ -0,0 +1,170 @@
+<?php
+/**
+ * REST API integration for the plugin.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Namespace for performance-lab REST API.
+ *
+ * @since n.e.x.t
+ * @var string
+ */
+const PERFLAB_REST_API_NAMESPACE = 'performance-lab/v1';
+
+/**
+ * Route for activating plugin/feature.
+ *
+ * Note the `:activate` art of the endpoint follows Google's guidance in AIP-136 for the use of the POST method in a way
+ * that does not strictly follow the standard usage.
+ *
+ * @since n.e.x.t
+ * @link https://google.aip.dev/136
+ * @var string
+ */
+const PERFLAB_FEATURES_ACTIVATE_ROUTE = '/features/(?P<slug>[a-z0-9_-]+):activate';
+
+/**
+ * Route for fetching plugin/feature information.
+ *
+ * @since n.e.x.t
+ * @var string
+ */
+const PERFLAB_FEATURES_INFORMATION_ROUTE = '/features/(?P<slug>[a-z0-9_-]+)';
+
+/**
+ * Registers endpoint for performance-lab REST API.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+function perflab_register_endpoint(): void {
+	register_rest_route(
+		PERFLAB_REST_API_NAMESPACE,
+		PERFLAB_FEATURES_ACTIVATE_ROUTE,
+		array(
+			'methods'             => 'POST',
+			'args'                => array(
+				'slug' => array(
+					'type'              => 'string',
+					'description'       => __( 'Plugin slug of plugin/feature that needs to be activated.', 'performance-lab' ),
+					'required'          => true,
+					'validate_callback' => 'perflab_validate_slug_endpoint_arg',
+				),
+			),
+			'callback'            => 'perflab_handle_feature_activation',
+			'permission_callback' => static function () {
+				if ( current_user_can( 'install_plugins' ) ) {
+					return true;
+				}
+
+				return new WP_Error( 'cannot_install_plugin', __( 'Sorry, you are not allowed to install plugins on this site.', 'performance-lab' ) );
```

Comment:
Checking this is not enough. For example, if the plugin you're sending a request for is already activated, we'd need to check for `activate_plugin` instead. At the bare minimum, we would need to check for both, but that would still lead to problems, e.g. for sites that disable writing to the filesystem (which means `install_plugins` will be `false` for everyone).

Therefore e need to implement this check more granularly, taking into account the given `slug`:
* If the plugin is not installed, check for `install_plugins` and `activate_plugins`.
    * Ideally the latter should instead use `activate_plugin` with the plugin's basename, but we don't know that unless the plugin is installed.
* If the plugin is installed, check for `activate_plugin` with the plugin's `plugin_basename( $file )` value.

Reply:
Note that `perflab_install_and_activate_plugin()` already has capability checks for `install_plugins` and `activate_plugin`, so adding the these granular permission callbacks here would be redundant, wouldn't they? 

At most I think this `permission_callback` should add an `current_user_can( 'activate_plugins' )` check:

```suggestion
				if ( current_user_can( 'install_plugins' ) && current_user_can( 'activate_plugins' ) ) {
					return true;
				}

				return new WP_Error( 'cannot_activate', __( 'Sorry, you are not allowed to activate this feature.', 'performance-lab' ) );
```

Reply:
To some degree it would be redundant. But if we don't want to have any redundancy, we need to _only_ check for `activate_plugins` here, not `install_plugins` - since `activate_plugins` is the only capability that you need to perform this action in any case. `install_plugins` is only relevant if the plugin is not installed yet, and we wouldn't want to prevent people from activating an already installed plugin only because they cannot install.

Reply:
Sounds good, so:

```suggestion
				// Important: The endpoint calls perflab_install_and_activate_plugin() which does more granular capability checks.
				if ( current_user_can( 'activate_plugins' ) ) {
					return true;
				}

				return new WP_Error( 'cannot_activate', __( 'Sorry, you are not allowed to activate this feature.', 'performance-lab' ) );
```